modules/fx/arrow-fx-kotlinx-coroutines/src/test/kotlin/arrow/integrations/kotlinx/CoroutinesIntegrationTest.kt

```diff
@@ -0,0 +1,166 @@
+package arrow.integrations.kotlinx
+
+import arrow.core.Either
+import arrow.core.None
+import arrow.core.right
+import arrow.core.some
+import arrow.fx.IO
+import arrow.fx.IOOf
+import arrow.fx.extensions.fx
+import arrow.fx.extensions.io.bracket.guaranteeCase
+import arrow.fx.typeclasses.ExitCase
+import arrow.fx.typeclasses.milliseconds
+import arrow.fx.typeclasses.seconds
+import arrow.test.UnitSpec
+import arrow.test.generators.throwable
+import io.kotlintest.fail
+import io.kotlintest.properties.Gen
+import io.kotlintest.properties.forAll
+import io.kotlintest.shouldBe
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.ObsoleteCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.newSingleThreadContext
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.TestCoroutineExceptionHandler
+import kotlinx.coroutines.test.TestCoroutineScope
+
+@ObsoleteCoroutinesApi
+@Suppress("IMPLICIT_NOTHING_AS_TYPE_PARAMETER")
+@UseExperimental(ExperimentalCoroutinesApi::class)
+class CoroutinesIntegrationTest : UnitSpec() {
+
+  private val other = newSingleThreadContext("other")
+  private val all = newSingleThreadContext("all")
+
+  init {
+    // --------------- suspendCancellable ---------------
+
+    "suspendedCancellable should throw" {
+      forAll(Gen.throwable()) { e ->
+        val ceh = TestCoroutineExceptionHandler()
+        val scope = TestCoroutineScope(ceh + TestCoroutineDispatcher())
+
+        scope.launch {
+          IO { throw e }.suspendCancellable()
+        }
+        // suspendCancellableCoroutine re throws the exception so we need to compare the instance
+        ceh.uncaughtExceptions[0]::class == e::class
```

Comment:
We should be able to perform instance checking here no?

Since the throwables are constructed by the `Gen` so the `stacktrace` etc is also constructed by the `Gen`, this happens at time of construction and not at the time of throwing IIRC so instance check should apply.

Reply:
That's what I though, but for some reason I'm getting the following:

> Expected --> java.lang.IllegalArgumentException#1073680203
Actual     --> java.lang.IllegalArgumentException#686013151

I just noticed that they're indeed copying the exception and rethrowing it, so I'll refer to this in the code 🤷‍♂️